### DISCUSSION TITLE
DNS server inside wsl2 is not working for windows

### DISCUSSION THREAD

**Discussion by feng-yifan** [microsoft/WSL#4592] [General]
I just installed a dnsmasq inside wsl2, and bind the windows' dns to its address, and I just found that the nslookup is working well but ping and curl is not working.
What should I do to fix it?

Here is the command and its result:
```
C:\Users\worker>nslookup gitlab.vm
server:  UnKnown
Address:  172.25.119.218

name:    gitlab.vm
Address:  192.168.1.201

C:\Users\worker>curl a.vm
curl: (6) Could not resolve host: gitlab.vm
```

**Comment by Biswa96** [comment]
Is curl running in WSL? The prompt text suggests its Command Prompt.

**Comment by feng-yifan** [comment]
> Is curl running in WSL? The prompt text suggests its Command Prompt.

No, both of the commands are running in windows terminal.
I also tried to use `nc` to test if the port is connectable and it works well.
```
C:\Users\worker>nc -vuz 172.25.119.218 53
218.119.25.172.in-addr.arpa [172.25.119.218] 53 (domain) open
```
```
root@workstation:~# nc -vuz 172.25.119.218 53
Connection to 172.25.119.218 53 port [udp/domain] succeeded!
```
Also `nslookup` givens the right answer.
My wsl2's ip address is *172.25.119.218* and its nameserver is *172.25.112.1* which I think is the host's address. Then I set the host's nameserver to *172.25.119.218*, so than I thought no matter where a request comes from it should be solved by the dnsmasq inside wsl2.

**Comment by valeryan** [comment]
I have been trying to get DnsMasq working in WSL2 as well. dnsmaq is running on the guest system and can resolve names just fine but I can't figure out how to make windows able to access that information. Setting 127.0.0.1 as the dns server for my network adapter doesn't work and using the ip address of the guest doesn't work but instead of just not working it times out.

It would be an incredible tool to have dnsmasq work for windows.

**Comment by valeryan** [comment]
I have done experiments with a few different setups, none of which work.

In WSL I have dnsmasq service running and listening on addresses (0.0.0.0).

I can use nslookup in bash to query all dns records for my local addresses like *.test and I get:
```
Server:         127.0.0.1
Address:        127.0.0.1#53

Name:   localhost.test
Address: 127.0.0.1
```

Now moving into windows without modifying the network adapter and using nslookup

```
nslookup localhost.test 127.0.0.1
Server:  UnKnown
Address:  127.0.0.1

*** UnKnown can't find localhost.test: No response from server
```
```
nslookup localhost.test 172.24.81.0 #Ip address returned from ifconfig in bash
Server:  UnKnown
Address:  172.24.81.0

Name:    localhost.test
Address:  127.0.0.1
```
It can also resolve internet addresses
```
nslookup google.com 127.24.81.0
Server:  UnKnown
Address:  172.24.81.0

Non-authoritative answer:
Name:    google.com
Addresses:  2607:f8b0:4009:807::200e
          172.217.8.206
```
```
nslookup localhost.test 172.24.80.1 #ip of hyper v bridge
Server:  DESKTOP-L4M323O.mshome.net
Address:  172.24.80.1

*** DESKTOP-L4M323O.mshome.net can't find google: Non-existent domain
```

Now if I modify the network adapter to point to 172.24.81.0 for dns, you would think that would work. However, Windows instead reports it has no internet and does not resolve any addresses. 

There has to be a way to make it work. My theory is that it causing an endless loop.

**Comment by feng-yifan** [comment]
> I have done experiments with a few different setups, none of which work.
> 
> In WSL I have dnsmasq service running and listening on addresses (0.0.0.0).
> 
> I can use nslookup in bash to query all dns records for my local addresses like *.test and I get:
> 
> ```
> Server:         127.0.0.1
> Address:        127.0.0.1#53
> 
> Name:   localhost.test
> Address: 127.0.0.1
> ```
> 
> Now moving into windows without modifying the network adapter and using nslookup
> 
> ```
> nslookup localhost.test 127.0.0.1
> Server:  UnKnown
> Address:  127.0.0.1
> 
> *** UnKnown can't find localhost.test: No response from server
> ```
> 
> ```
> nslookup localhost.test 172.24.81.0 #Ip address returned from ifconfig in bash
> Server:  UnKnown
> Address:  172.24.81.0
> 
> Name:    localhost.test
> Address:  127.0.0.1
> ```
> 
> It can also resolve internet addresses
> 
> ```
> nslookup google.com 127.24.81.0
> Server:  UnKnown
> Address:  172.24.81.0
> 
> Non-authoritative answer:
> Name:    google.com
> Addresses:  2607:f8b0:4009:807::200e
>           172.217.8.206
> ```
> 
> ```
> nslookup localhost.test 172.24.80.1 #ip of hyper v bridge
> Server:  DESKTOP-L4M323O.mshome.net
> Address:  172.24.80.1
> 
> *** DESKTOP-L4M323O.mshome.net can't find google: Non-existent domain
> ```
> 
> Now if I modify the network adapter to point to 172.24.81.0 for dns, you would think that would work. However, Windows instead reports it has no internet and does not resolve any addresses.
> 
> There has to be a way to make it work. My theory is that it causing an endless loop.

On my case, `nslookup` command is working well you can get the right dns result, but if you try `curl` it dosen't work, which really confused me.
Now I just use dnsmasq inside the wsl to tell services how to visit each other, until I find my final solution.

Here is my temporary solution:
- add all custom domains into the windows hosts file
- add a script inside wsl which can replace the ip address inside hosts file with wsl's ip address 
- add a task in windows so that everytime you login windows that script will run automatically 

Here is my hosts file
```
hosts
172.19.191.22		gitlab.wsl
172.19.191.22		dev.wsl
```
Here is my scripts
```
autostart.sh
echo "password" | sudo -S service ssh start &&      \
        sudo -S service docker start && \
        php /var/shell/update-hosts.php
```
```
update-hosts.php
<?php
$net_name = 'eth0';
$hosts_path = '/mnt/c/Windows/System32/drivers/etc/hosts';
exec("ifconfig",$ifconfig);
foreach($ifconfig as $key => $result) {
        if (substr($result,0,4) === $net_name) break;
}
$ip_addr = explode(' ',$ifconfig[++$key])[9];
$rules = [
        [-5,4,'.wsl',$ip_addr],
        [-6,5,'t.wsl','127.0.0.1'],
];
if(filter_var($ip_addr, FILTER_VALIDATE_IP)) {
        $hosts = file_get_contents($hosts_path);
        $new_hosts = $hosts = explode("\n",$hosts);
        foreach ($hosts as $key => $host) {
                foreach ($rules as $rule) {
                        if (!empty($host) && substr($host,0,1) !=='#' && substr($host,$rule[0],$rule[1]) === $rule[2]) {
                                $host_data = explode("\t",$host);
                                $host_data[0] = $rule[3];
                                $new_hosts[$key] = implode("\t",$host_data);
                                echo "╔═ ".$host.PHP_EOL."╚═▶".$new_hosts[$key].PHP_EOL;
                        }
                }
        }
        file_put_contents($hosts_path,implode("\n",$new_hosts));
} else {
        echo "+-------------------------+".PHP_EOL;
        echo "|     Invalid Ip Addr     |".PHP_EOL;
        echo "+-------------------------+".PHP_EOL;
}
```
And my windows task action
```
windows task setting
C:\Windows\System32\bash.exe -c "/var/shell/autostart.sh"
```

**Comment by valeryan** [comment]
That does not provide the ability to have wildcard domains inside windows. With dnsmasq I can configure *.tld to resolve to localhost. Windows host file is very limited and dnsmasq could replace it's archaic system and make windows development more than just an experiment. I maintain Valet for WSL and dnsmasq is the final piece I need for development to be viable with WSL. The only other option is using acrylic dns proxy in windows. The native dnsmasq works way better for Valet.

**Comment by iljhli** [comment]
I'm trying to make this happen as well. I bought a Surface recently, and almost wish I'd have bought a MacBook Pro, but I'm hopeful WSL 2 will make Windows development less of a pain. The following are my configurations, tests, and findings.  There seems to be a loop, and I've been trying various things with configurations that I may have tunnel vision and can't see if I'm missing something simple.

Ethernet Adapter:
```
IP 10.0.0.135
IPv4 DNS Pref 172.18.209.28 (eth0 iface of WSL Distro)
IPv4 DNS Sec 8.8.8.8 (Google)
IPv6 DNS Pref fe80::215:5dff:febb:4c6c (eth0 iface of WSL Distro)
IPv6 DNS Sec 2001:4860:4860::8888 (Google)
```
/etc/resolve.conf:
```
nameserver 127.0.0.1
nameserver 8.8.8.8
```
/etc/dnsmasq.conf:
```
address=/.test/127.0.0.1
```
In the following images Ethernet 2 is network adapter, vEthernet WSL, *any is from any iface in linux distro
 
nslookup google.com (from WSL distro):
```

Server:         127.0.0.1
Address:        127.0.0.1#53

Non-authoritative answer:
Name:   google.com
Address: 172.217.9.206
Name:   google.com
Address: 2607:f8b0:4004:806::200e
```
![image](https://user-images.githubusercontent.com/58231895/69801185-b2c84700-11a4-11ea-86e6-953a99cd0d2c.png)

nslookup example.test (from WSL distro):
```
Server:         127.0.0.1
Address:        127.0.0.1#53

Name:   example.test
Address: 127.0.0.1
```

![image](https://user-images.githubusercontent.com/58231895/69801396-24a09080-11a5-11ea-8d07-e31e00889d96.png)

nslookup google.com (from Windows):
```
DNS request timed out.
    timeout was 2 seconds.
Server:  UnKnown
Address:  fe80::215:5dff:febb:4c6c

DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
*** Request to UnKnown timed-out
```
![image](https://user-images.githubusercontent.com/58231895/69801627-a8f31380-11a5-11ea-8e43-e674527042bc.png)

nslookup example.test yields the same result as google.com from windows.

nslookup google.com 172.18.209.28 (from Windows using eth0 iface):
```
Server:  UnKnown
Address:  172.18.209.28

Non-authoritative answer:
Name:    google.com
Addresses:  2607:f8b0:4004:815::200e
          172.217.9.206
```

![image](https://user-images.githubusercontent.com/58231895/69801899-4ea68280-11a6-11ea-9b63-31984f5326f6.png)

nslookup example.test 172.18.209.28 (from Windows using eth0 iface):
```
Server:  UnKnown
Address:  172.18.209.28

Name:    example.test
Address:  127.0.0.1
```

![image](https://user-images.githubusercontent.com/58231895/69802064-b78dfa80-11a6-11ea-8401-3284ab58b9c0.png)

Using IPv6 iface address in place of the last IPv4 yields the same results.

**Comment by brad-jones** [comment]
I am having a very similar issue but just running dnsmasq inside an Alpine Linux Hyper-V VM.

My workaround for now has been to use Acrylic to forward all queries to the dnsmasq server. It works but obviously isn't ideal.

When ever I reconfigure my network interface(s) to talk directly to the dnsmasq server then name resolution fails.

**Comment by muodov** [comment]
The recent update with WSL2 broke this for me. Previously I've been using Acrylic proxy for a "wildcard" alternative to hosts file, and now it seems like there is literally no way of doing that on windows. It's incredibly frustrating because I'm developing a proxy and this functionality is essential.

**Comment by haggen** [comment]
I managed to get it working just now. I'm using Debian. 

1. Installed dnsmasq via aptitude, and only added my desired records for localhost subdomains `address=/.localhost/127.0.0.1`.
2. Edited `/etc/resolve.conf` and added `nameserver 127.0.0.1` before the existing `nameserver` line.
3. Back to Windows, I simply edited my WSL virtual interface and added my guest Linux machine as the DNS. Left my primary interface untouched.

It works! :)

**Update.** Domain resolution became insanely slow so as a workaround I added upstream servers to my dnsmasq configuration:

```
server 1.1.1.1
server 1.0.0.1
```

That did the trick.

**Update 2.** It didn't. After a reboot it's still working but DNS resolution is slow again. Like many several seconds slow.

**Update 3.** I set my preferred DNS in my primary network interface on Windows and followed these [steps to change interface binding order](https://serverfault.com/a/805455) and it seems to have corrected the issue of slow DNS resolution. Before, running a nslookup to `google.com` would give several timeouts before returning a match. Now it's instantaneous.

**Comment by jerivas** [comment]
Also trying to set up dnsmasq to handle *.test domains internally. Here' what I've done (Ubuntu 20.04):

1. Add `nameserver 127.0.0.1` to the top of `/etc/resolve.conf`
2. Make sure `dnsmasq` is running with `sudo service dnsmasq start`
3. Added wildcard settings to `/etc/dnsmasq.conf`: `address=/.test/127.0.0.1`
4. Added external DNS servers to `/etc/dnsmasq.conf`: `server=1.1.1.1` (Cloudflare)

At this point if I run `dig` from Ubuntu and it seems like *.test resolve correctly:

```bash
dig facebook.com +short
157.240.14.35
dig example.test +short
127.0.0.1
```
Sadly in Windows land `nslookup` always fails to find .test domains. I've tried setting my main network adapter IPv4 settings to use 127.0.0.1 as name server, as well as the local IP that Ubuntu reports when running `ifconfig`

**Comment by feng-yifan** [comment]
> Also trying to set up dnsmasq to handle *.test domains internally. Here' what I've done (Ubuntu 20.04):
> 
> 1. Add `nameserver 127.0.0.1` to the top of `/etc/resolve.conf`
> 2. Make sure `dnsmasq` is running with `sudo service dnsmasq start`
> 3. Added wildcard settings to `/etc/dnsmasq.conf`: `address=/.test/127.0.0.1`
> 4. Added external DNS servers to `/etc/dnsmasq.conf`: `server=1.1.1.1` (Cloudflare)
> 
> At this point if I run `dig` from Ubuntu and it seems like *.test resolve correctly:
> 
> ```shell
> dig facebook.com +short
> 157.240.14.35
> dig example.test +short
> 127.0.0.1
> ```
> 
> Sadly in Windows land `nslookup` always fails to find .test domains. I've tried setting my main network adapter IPv4 settings to use 127.0.0.1 as name server, as well as the local IP that Ubuntu reports when running `ifconfig`

I did the exactly same steps with you and got the same result.
I tried use `nc -luv 0.0.0.0 53` to watch if there is any dns request, and it has no respond.
I think this will take a long while to be solved.

**Comment by haggen** [comment]
> Sadly in Windows land `nslookup` always fails to find .test domains. I've tried setting my main network adapter IPv4 settings to use 127.0.0.1 as name server, as well as the local IP that Ubuntu reports when running `ifconfig`

Don't set the DNS of your main network. Set the DNS of your WSL network with the IP of your Linux VM. That's what did the trick for me. I also had to set a DNS in my main network but this time pointed to your preferred DNS service, like Google or CloudFlare.

**Comment by therealkenc** [comment]
This issue had no repro, but the spirit was answered well by @haggen [above](https://github.com/microsoft/WSL/issues/4592#issuecomment-642014948). Whether bouncing every DNS query in Windows off a WSL instance is a good idea can be left to the beholder. Having more than one thing listening on port `53` is going to be (at best) precarious, but is possible. For its part, Linux dnsmasq is behaving correctly. Feel free to continue the discussion in the newly-minted discussion forum though. 

For an alternate approach to accomplishing a similar end result, some people _may_ want to consider [`avahi-daemon(8)`](http://manpages.ubuntu.com/manpages/bionic/man8/avahi-daemon.8.html).
